### ISSUE TITLE
How to get context?

### ISSUE THREAD

**Issue by hack1234567** [JRC1995/Abstractive-Summarization#3]
plz help me to understand how to select the next word to be predicted and context for it.
thank you

**Comment by JRC1995** [comment]
This program is based on the encoder-decoder model. I am not sure exactly if you want to know the whole thing or what I am exactly doing in the decoder part. The encoder encodes the input sentences - usually uses the context of the previous words to encode the current word. A bi-directional encoder encodes by taking context from both sides. That is, once it encodes a word by taking the context of past encoded words, and once it encodes a word by taking the context of future encoded words. Then it combines the two versions of encoding into one by some fashion (addition, average or something). As a result, for each input word (or word-vector) you will end up with its encoded version.

The prediction happens in the decoder part. The predicted word in case of summarization should depend on the input (in this case the encoded version of the input) which is being summarized, and the context of the previous predicted words. When predicting the first word there is no previous predicted word. But we usually a word vector which may be symbolized <SOS> (representing start of sentence) as the first word to provide context for the first prediction. This is done to use to same formula for predicting each of the words. In addition to the previous context provided by the previous words, we also take the context from the encoded input (that which is to be summarized). Attention mechanism is used to select a relevant portion of encoded input to serve as a context. Both the context from previous words, and the encoded input are combined and used as a hidden state for the decoder RNN to finally predict the next word.  

The dataset I used comes with example inputs and their corresponding summaries (outputs that we want the algorithm to predict given the inputs). The summaries from the dataset can be considered as target outputs...and what the model actually produces are the predictions. The model's prediction is compared with the target output, and then some learning algorithm is used to make the machine predict summarizations similar to the target outputs.

**Comment by hack1234567** [comment]
Thank you for taking time to answer.
Also, does using encoder decoder with attention eliminate the need  to check grammar..... for eg john dies vs john died vs john dead... will this model be able to distinguish between dies and died? 
what happens when we have to get output like : "the book is great" how to identify "the" and "is"?
thank you.

**Comment by JRC1995** [comment]
> john dies vs john died vs john dead... will this model be able to distinguish between dies and died?
Yes, the model can distinguish (superficially, at least) 'died' and 'dead, because normally 'died' and 'dead' will be represented by different _word vectors_ when they are fed to the model. But, that doesn't necessarily mean it will understand their grammatical meaning and significance. 

What the model learns to predict depends on the dataset. For example, the model may learn from the data that it is not very likely that 'dead' will follow after a noun....and 'dead' is more likely to follow after 'is' or something like that. The model may learn that if the previous words in the sentence were in past tense, then from the context of the past words 'john dead' is more likely that 'john dies'. If the input mention 'john dies' then the model may learn that the output is more likely to be 'john dies'...and so on.

Similarly, the model should learn that 'the' often starts at the beginning at the sentence or before a noun like word, and almost never after a noun, where as 'is' often start after a noun. So the model will give higher probability to 'is' than 'the' given that the previous words are "the book".

The model basically start to learn the patterns of our usage of words from the huge chunk of data it is fed. Thus, if the data has poor quality grammar, the model output will probably also be of poor quality grammar. 

Other machine learning models are mostly similar in this respect.

Does it totally eliminate the need to check grammar? Well, if you do check the grammar, or write some program at the end to enhance the grammar of the output you may get better result. Abstractive summarization is a work in progress. It's not yet perfect. And in practice, the grammar of the output may not be perfect either even if trained from high quality dataset. I am not sure about the exact quality of the output of the current state of the art abstractive summarization programs. The target is to eliminate the need to write specific grammatical rules and make the model learn for itself.

**Comment by hack1234567** [comment]
thanks a lot for the detailed explanation. Please bear with me for  2 more questions.
Have you implemented noun identification? Is it possible for me to use any other big dataset(say news article) and say train it little more time using the same code and expect similar or slightly higher accuracy in prediction(if its a big dataset).
thank you

**Comment by JRC1995** [comment]
I haven't implemented noun identification. But I have used NLTK library for POS (part-of-speech) tagging i.e tagging words as NN (singular noun), VB (verb base form) and so on (http://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html).

You may find some open source code for POS tagging somewhere. 

> Is it possible for me to use any other big dataset(say news article) and say train it little more time using the same code and expect similar or slightly higher accuracy in prediction(if its a big dataset).

It's possible. 

If you plan to use this code, I will recommend using mini-batching which I didn't use in this code. 

I used mini batching in my later codes (like: https://github.com/JRC1995/Machine-Translation-Transformers) but those are mostly different models. 
 
You may try using this (https://github.com/JRC1995/Machine-Translation-Transformers) code for summarization too. It is also essentially an encoder-decoder model but without RNNs and sequential processing. 
 
Also to objectively compare the quality of an output, evaluation metrics like ROGUE\BLEU\METEOR etc. can be used.

Also note, I haven't used the validation and test set. Validation accuracy and the evaluation of predictions of validation input is important to understand where to stop the training or which state of the model to use for prediction. And, test accuracy is usually the best reflection of the model's performance.

**Comment by hack1234567** [comment]
I will try with mini-batching. Thanks a lot for your help
